From c327bb81044c7f4f6db3a3b6501b63bb2db9d4ee Mon Sep 17 00:00:00 2001
From: Frank Ramsay <framsay@redhat.com>
Date: Thu, 13 Nov 2014 18:29:57 -0500
Subject: [x86] uv_bau: Avoid NULL pointer reference in ptc_seq_show

Message-id: <1415903397-11300-3-git-send-email-framsay@redhat.com>
Patchwork-id: 99960
O-Subject: [RHEL 7.1 PATCH 2/2] x86: UV BAU: Avoid NULL pointer reference in ptc_seq_show
Bugzilla: 1161183
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>

Commit: fa2a79ce6aef5de35a4d50487da35deb6b634944
Author: James Custer <jcuster@sgi.com>
Date: 2014-11-02 18:16:38 (GMT)

 In init_per_cpu(), when get_cpu_topology() fails, init_per_cpu_tunables() is
 not called afterwards. This means that bau_control->statp is NULL. If a user
 then reads /proc/sgi_uv/ptc_statistics ptc_seq_show() references a NULL
 pointer. Therefore, since uv_bau_init calls set_bau_off when init_per_cpu()
 fails, we add code that detects when the bau is off in ptc_seq_show() to
 avoid referencing a NULL pointer. Signed-off-by: James Custer
 <jcuster@sgi.com> Cc: Russ Anderson <rja@sgi.com> Link:
 http://lkml.kernel.org/r/1414952199-185319-2-git-send-email-jcuster@sgi.com
 Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/platform/uv/tlb_uv.c b/arch/x86/platform/uv/tlb_uv.c
index efe4d72..a6b5aff 100644
--- a/arch/x86/platform/uv/tlb_uv.c
+++ b/arch/x86/platform/uv/tlb_uv.c
@@ -1367,6 +1367,10 @@ static int ptc_seq_show(struct seq_file *file, void *data)
  }
  if (cpu < num_possible_cpus() && cpu_online(cpu)) {
   bcp = &per_cpu(bau_control, cpu);
+  if (bcp->nobau) {
+   seq_printf(file, "cpu %d bau disabled\n", cpu);
+   return 0;
+  }
   stat = bcp->statp;
   /* source side statistics */
   seq_printf(file,
-- 
1.7.1